### PR TITLE
Overall refactor of extractors

### PR DESCRIPTION
**Pull request by jpihl**

### CODE REVIEW

File: src/petro/extractor/avc_sample_access.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    public:
+
+        void read(uint32_t size_length, const uint8_t* data, uint32_t size)
+        {
+            m_nalus.clear();
+            m_nalu_sizes.clear();
+
+            while(size != 0)
+            {
+                auto nalu_size = detail::read_bytes(size_length, data);
```
Comment (mortenvp) [inline src/petro/extractor/avc_sample_access.hpp]: Does this belong in detail.. I guess it can just be in the extractor namespace

File: src/petro/extractor/avc_sample_extractor.hpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#include "sample_extractor_layer.hpp"
+#include "avc_track_layer.hpp"
+#include "parser_layer.hpp"
+#include "memory_mapped_file_layer.hpp"
+
+namespace petro
+{
+namespace extractor
+{
+    /// stack for extracting avc samples
```
Comment (mortenvp) [inline src/petro/extractor/avc_sample_extractor.hpp]: Put a link to an example the uses the extractor (if there is one).

File: src/petro/extractor/avc_track_layer.hpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+#include <memory>
+
+#include "../box/box.hpp"
+#include "../box/avcc.hpp"
+
+namespace petro
+{
+namespace extractor
+{
+    template<class Super>
```
Comment (mortenvp) [inline src/petro/extractor/avc_track_layer.hpp]: Purpose of the class

File: src/petro/extractor/looper_layer.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+#pragma once
+
+#include <cstdint>
+
+namespace petro
+{
+namespace extractor
+{
+    /// Helper layer allowing extractors to be looping.
```
Comment (mortenvp) [inline src/petro/extractor/looper_layer.hpp]: Awesome comment!!

Comment (jpihl) [inline src/petro/extractor/looper_layer.hpp]: thx

File: src/petro/extractor/parser_layer.hpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+#include "../box/stts.hpp"
+#include "../box/stsz.hpp"
+#include "../box/stbl.hpp"
+#include "../box/root.hpp"
+
+namespace petro
+{
+namespace extractor
+{
+    template<class Super>
```
Comment (mortenvp) [inline src/petro/extractor/parser_layer.hpp]: Missing comment..

**Comment by petya2164** [comment]
@jpihl I have to say that the overall readability of the extractors is greatly diminished by the application of mixin layers! It is quite hard to follow what is happening and where. The code was very clear and simple before this change. You might have removed a bit of duplicated code, but I really think this is an overkill. 
We should be very conservative about applying the "magical" mixin layers, otherwise we increase the code complexity for no good reason :( It was also an overkill in fifi :(

**Comment by jpihl** [comment]
I'm sorry that you feel that way. It was the only way I was able to reuse the code for the two extractors, while keeping the code both testable and extendable.
I started out with the composite pattern, but that approach made it necessary to forward a lot of calls to the internal extractor.
Can you elaborate on what you think is difficult to understand? I can maybe add some more comments to make it more clear?

**Comment by petya2164** [comment]
It is not a feeling, it is an observation! In the previous version, we had 2 classes to implement the extractors, but now we have 17 for the same thing. That is just horrible from a readability and complexity perspective :( Objectively, you have to agree with that assessment...
Of course, we are trying to balance readability, complexity, code duplication and testability everywhere in our codebase, and it is not easy to find a reasonable balance and sometimes we just go overboard. If testability is your concern, then we should really find a simpler solution, because it is just wrong to move each function to its own layer :(
But this discussion is too late for this project, so I am just asking to be very conservative when you consider mixin layers in the future.

**Comment by jpihl** [comment]
Yes the new approach is more complex.

But other than that I have to disagree. It's not the same thing. The current version is not only much more testable, but the added reusability and extendability allows the creation of extractors for the [many other codecs](http://www.mp4ra.org/codecs.html) contained in mp4 files.
I haven't moved each *function* to it's own layer, I've been very keen on moving *responsibilities* to different layers. Of course this will have an impact on the overview and that might affect how easy it is for people to use the library. I think a good way to solve that is using the pimpl Idiom, which has the added benefit of firewalling the dependency on boost.
I don't think a discussion is too late, but a closed PR might not be the place right place to have this discussion, let's move it to slack.
